### PR TITLE
modularize split_n_preprocessor file

### PR DESCRIPTION
**Pull request by WeilinHan8**
Hi,
Please kindly review the modularized split_n_preprocessor file and other files in src folder.

### CODE REVIEW

File: scripts/split_n_preprocess.py [hunk 1]
```diff
@@ -6,90 +6,44 @@
...
-
+sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
+from src.split_train_test import split_train_test_data, validate_split_data
+from src.clean_data import write_data
+from src.preprocessor import create_save_preprocessor
 
 @click.command()
-@click.option('--raw-data', type=str, help="Path to raw data")
+@click.option('--cleaned-data', type=str, help="Path to cleaned data")
+@click.option('--train-data-size', type=str, help="Size of the dataset to include in the train split")
```
Comment (ttimbers) [inline scripts/split_n_preprocess.py:17]: Can you update the `Makefile` the the necessary changes to run it with this update to the script (that we can not edit the train data size)?

File: src/preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import os
+import pickle
+from sklearn.preprocessing import StandardScaler
+from sklearn.compose import make_column_transformer, make_column_selector
+
+def create_save_preprocessor(train_data, test_data, data_to, preprocessor_to):
```
Comment (ttimbers) [inline src/preprocessor.py:6]: This function does too many things. I suggest we break this into two functions:

1. `create_preprocessor` - this function just sets up the preprocessor, it will likely be used just once. We create a function so we can test that our code works as expected.
2. `write_model` - this function can be used to save any model object, the preprocessor here, but the fit model later too. This is similar to our `write_data` function.

And we can use our `write_data` function to do the saving of the scaled data?

File: src/split_train_test.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        raise TypeError("The cleaned_data must be a pandas DataFrame.")
+
+    # Ensure train_data_size is a valid proportion
+    if not (0 < train_data_size < 1):
+        raise ValueError("train_data_size must be a float between 0 and 1.")
+
+    # Handle stratify_by
+    if stratify_by:
+        if stratify_by not in cleaned_data.columns:
+            raise KeyError(f"The column '{stratify_by}' does not exist in the cleaned_data.")
```
Comment (ttimbers) [inline src/split_train_test.py:21]: Can you double check that `sklearn` does not already handle these well. If it does, then we can remove them.

File: src/split_train_test.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    
+    # create the split
+    data_train, data_test = train_test_split(
+        cleaned_data, train_size=train_data_size, stratify=stratify_by #cleaned_data["diagnosis"]
```
Comment (ttimbers) [inline src/split_train_test.py:25]: remove the comment.

File: src/split_train_test.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    
+    return data_train, data_test
+
+def validate_split_data(data_train, data_test):
+    """Test split datasets using Deepchecks"""
+
+    # test train test split dataframes
+    # prepare the train test datasets to Deepchecks Dataset format
+    data_train = Dataset(data_train,features=data_train.columns[1:],label=data_train.columns[0])
+    data_test = Dataset(data_test,features=data_test.columns[1:],label=data_test.columns[0])
```
Comment (ttimbers) [inline src/split_train_test.py:36]: Please change from numerical subsetting to using column names/labels. The function can be made flexible to take in column names/labels. Numerical subsetting is very brittle and error-prone.

File: src/split_train_test.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    # Datasets Size Comparison Check
+    check_instance = (
+    DatasetsSizeComparison()
+    .add_condition_train_dataset_greater_or_equal_test()
+    .add_condition_test_train_size_ratio_greater_than(0.2)
```
Comment (ttimbers) [inline src/split_train_test.py:42]: What is this magic number for? Maybe add a comment to explain? Or should this also be a function input?

File: src/split_train_test.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    )  
+    data_size_comp = check_instance.run(data_train, data_test)
+    if not data_size_comp.passed_conditions():
+        raise ValueError("The train test data size ratio should be greater than 0.2")
```
Comment (ttimbers) [inline src/split_train_test.py:46]: related to the point above....

File: src/split_train_test.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            raise KeyError(f"The column '{stratify_by}' does not exist in the cleaned_data.")
+    
+    # create the split
+    data_train, data_test = train_test_split(
+        cleaned_data, train_size=train_data_size, stratify=stratify_by #cleaned_data["diagnosis"]
+    )
+    
+    return data_train, data_test
+
+def validate_split_data(data_train, data_test):
```
Comment (ttimbers) [inline src/split_train_test.py:30]: I feel like this function should take in inputs for all these thresholds. This would make the function more reusable across projects. Maybe the input could be a dictionary?

**Comment by ttimbers** [comment]
Can we also please add numpy-style docstrings to each function here? At least the ones in the `src`? This will help me review the code more easily because there is a place I can reference your design choices against your code. This would be particulary useful for `validate_split_data` so that I can easily understand which checks on the checklist this function is validating. Also, we should elaborate a little as to why we do these checks in English (like what is Feature Drift Check, and why check test and train data for Feature Drift Check)?

**Comment by WeilinHan8** [comment]
> Maybe we already chatted about the test data, but does it need to be so big? Can it be less columns and rows and still work?

One hundred observations is the minimum number of observations that deepChecks drift checks allowed.
